sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,128 @@
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_iot_hub_client_options const* options)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(iot_hub_hostname, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
+  AZ_PRECONDITION_NOT_NULL(options);
+
+  client->_internal.iot_hub_hostname = iot_hub_hostname;
+  client->_internal.device_id = device_id;
+  client->_internal.options.module_id = options->module_id;
+  client->_internal.options.user_agent = options->user_agent;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_hub_client_user_name_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_user_name,
+    az_span* out_mqtt_user_name)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+
+  const az_span* module_id = &(client->_internal.options.module_id);
+  const az_span* user_agent = &(client->_internal.options.user_agent);
+
+  if (az_span_capacity(mqtt_user_name)
+      < (int32_t)(
+            az_span_length(client->_internal.iot_hub_hostname) + sizeof(hub_client_forward_slash)
+            + az_span_length(client->_internal.device_id) + sizeof(hub_client_forward_slash)
+            + az_span_length(*module_id)
+            + (az_span_length(*module_id) > 0 ? sizeof(hub_client_forward_slash) : 0)
+            + az_span_length(*user_agent)
+            + (az_span_length(*user_agent) > 0 ? sizeof(hub_client_param_separator) : 0)
+            + az_span_length(hub_client_api_version) + sizeof(hub_client_null_terminate)))
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY;
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, client->_internal.iot_hub_hostname, out_mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_user_name, hub_client_forward_slash, out_mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append(*out_mqtt_user_name, client->_internal.device_id, out_mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_user_name, hub_client_forward_slash, out_mqtt_user_name));
+  if (az_span_length(*module_id) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_span_append(*out_mqtt_user_name, *module_id, out_mqtt_user_name));
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(*out_mqtt_user_name, hub_client_forward_slash, out_mqtt_user_name));
+  }
+  AZ_RETURN_IF_FAILED(
+      az_span_append(*out_mqtt_user_name, hub_client_api_version, out_mqtt_user_name));
+  if (az_span_length(*user_agent) > 0)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(*out_mqtt_user_name, hub_client_param_separator, out_mqtt_user_name));
+    AZ_RETURN_IF_FAILED(az_span_append(*out_mqtt_user_name, *user_agent, out_mqtt_user_name));
+  }
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_user_name, hub_client_null_terminate, out_mqtt_user_name));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_hub_client_id_get(
```

Comment:
Can we have some comments that show sample output of this impl?
Atleast the format that it will emit.

Reply:
I mean all the API's impl

---
In reply to: [392373089](https://github.com/Azure/azure-sdk-for-c/pull/455#discussion_r392373089) [](ancestors = 392373089)

Reply:
whoops yea will add to the README with the telemetry snippet.

Reply:
That will be great for customer, what about developers who are looking at the impl? It is important to have comments on *c file also.

---
In reply to: [392373835](https://github.com/Azure/azure-sdk-for-c/pull/455#discussion_r392373835) [](ancestors = 392373835)

Reply:
Would adding something like this be sufficient?
```C
// [Format without module id] {iothubhostname}/{device_id}/?api-version=2018-06-30
// [Format with module id] {iothubhostname}/{device_id}/{module_id}/?api-version=2018-06-30
```

Reply:
added above the functions in the `.c`'s and in the doxygen